### PR TITLE
Use separable filter for SSIM calculation to speed up.

### LINKED ISSUE #1850: Faster implementation of MS-SSIM
## 🚀 Feature

Current implementation impacts training speed significantly (~40% slowdown compared to pure L1)

## Motivation

MS-SSIM gives some awesome results, and it'd be nice to have an option to switch to a 
* lower accuracy and faster implementation such as https://link.springer.com/article/10.1007/s11554-010-0170-9
* more optimized implementation

## Pitch

Faster iteration is always great. Difference between workflow for a task which starts in morning and ends during lunch compared to one that ends in evening is huge.

## Alternatives

Reduce the scales in MS-SSIM

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by dodobyte**
#### Changes
Gaussian kernel is separable, the original implementation uses 2d convolution with 2d kernels which is very slow with large kernels.

Fixes #1850

### CODE REVIEW

**Comment by johnnv1** [comment]
@dodobyte can you provide a benchmark for this?

File: kornia/metrics/ssim.py [hunk 1]
```diff
@@ -75,15 +75,16 @@ def ssim(
...
     C2: float = (0.03 * max_val) ** 2
 
     # compute local mean per channel
-    mu1: torch.Tensor = filter2d(img1, kernel)
-    mu2: torch.Tensor = filter2d(img2, kernel)
+    mu1: torch.Tensor = filter2d_separable(img1, kernel_x, kernel_y)
+    mu2: torch.Tensor = filter2d_separable(img2, kernel_x, kernel_y)
 
     cropping_shape: List[int] = []
     if padding == 'valid':
```
Comment (johnnv1) [inline kornia/metrics/ssim.py:90]: needs to fix the h and w at this branch

Comment (johnnv1) [reply kornia/metrics/ssim.py:90]: btw, it doesn't have tests covering it too

Comment (dodobyte) [reply kornia/metrics/ssim.py:90]: hey, sorry I wasn't familiar with the practices. I tested it myself both to make sure that it produces the same result and see the speed difference. My test code is [here](https://gist.github.com/dodobyte/9a87bd68561df4548eebfb02ac99f702).

the benchmark for different kernels is the following, image resolutions are 1024x1024, time is averaged over 1000 runs.

| Kernel  | SSIM | SSIM-Sep. | Time | Time-Sep. |
| --- | --- | --- | --- | --- |
| 11  | 0.8567 | 0.8567  | 1.55 ms|1.22 ms|
| 21  | 0.8566 | 0.8566  | 3.62 ms|1.54 ms|
| 31  | 0.8566 | 0.8566  | 7.18 ms|1.86 ms|
| 51  | 0.8566 | 0.8566  | 17.98 ms|2.49 ms|
| 101  | 0.8566 | 0.8566  | 66.95 ms|4.06 ms|

Comment (dodobyte) [reply kornia/metrics/ssim.py:90]: by the way, I also adapted tensorflow's implementation and it's faster as it has one less (4) convolution calls.
I'll leave that code here as a reference as well as the benchmark.

| Kernel | Time | Time-Sep. | Time-TF |
| --- | --- | --- | --- |
| 11  | 1.55 ms|1.22 ms|0.91 ms|
| 21  | 3.62 ms|1.54 ms|1.16 ms|
| 31  | 7.18 ms|1.86 ms|1.41 ms|
| 51  | 17.98 ms|2.49 ms|1.92 ms|
| 101 | 66.95 ms|4.06 ms|3.19 ms|

```python
def ssim_tf(x, y, window=11, sigma=1.5, max_val=1.0):
	# prepare kernel
	kernel_x = get_gaussian_kernel1d(window, sigma).cuda()
	kernel_y = get_gaussian_kernel1d(window, sigma).cuda()

	c1 = (0.01 * max_val)**2
	c2 = (0.03 * max_val)**2

	# SSIM luminance measure is
	# (2 * mu_x * mu_y + c1) / (mu_x ** 2 + mu_y ** 2 + c1).
	mean0 = filter2d_separable(x, kernel_x, kernel_y)
	mean1 = filter2d_separable(y, kernel_x, kernel_y)
	num0 = mean0 * mean1 * 2.0
	den0 = mean0 ** 2 + mean1 ** 2
	luminance = (num0 + c1) / (den0 + c1)

	# SSIM contrast-structure measure is
	#   (2 * cov_{xy} + c2) / (cov_{xx} + cov_{yy} + c2).
	num1 = filter2d_separable(x * y, kernel_x, kernel_y) * 2.0
	den1 = filter2d_separable(x ** 2 + y ** 2, kernel_x, kernel_y)
	cs = (num1 - num0 + c2) / (den1 - den0 + c2)

	# SSIM score is the product of the luminance and contrast-structure measures.
	return luminance * cs
```

Comment (edgarriba) [reply kornia/metrics/ssim.py:90]: Great ! If tests are passing I’d move to this implementation

Comment (johnnv1) [reply kornia/metrics/ssim.py:90]: @dodobyte can you fix line 90 and add a test for it?

Comment (dodobyte) [reply kornia/metrics/ssim.py:90]: sure will do that now.

Comment (johnnv1) [reply kornia/metrics/ssim.py:89]: Oh, we don't have tests at all for this (related to #2231) 

But it should be like https://github.com/kornia/kornia/blob/master/test/test_metrics.py#L225

File: kornia/metrics/ssim.py [hunk 2]
```diff
@@ -75,19 +75,19 @@ def ssim(
         raise ValueError(f"img1 and img2 shapes must be the same. Got: {img1.shape} and {img2.shape}")
 
     # prepare kernel
-    kernel: torch.Tensor = get_gaussian_kernel2d((window_size, window_size), (1.5, 1.5))
+    kernel: torch.Tensor = get_gaussian_kernel1d(window_size, 1.5)
```
Comment (johnnv1) [inline kornia/metrics/ssim.py:78]: as suggested by @gau-nernst
```suggestion
    kernel: torch.Tensor = get_gaussian_kernel1d(window_size, 1.5, device=img1.device, dtype=img1.dtype)
```

**Review by johnnv1 — APPROVED** [review]
LGTM

File: kornia/metrics/ssim.py (hunk 1 above)
Comment (johnnv1) [reply kornia/metrics/ssim.py:89]: This can be done into a subsequent PR 😄

**Comment by johnnv1** [comment]
just for the record, results with [torch benchmark](https://pytorch.org/tutorials/recipes/recipes/benchmark.html)

arguments in the orders: `device`, `batch size`, `resolution`, `kernel` -- x and y generated with `torch.rand(batchsize, 3, res, res)`

```output
[--------------------------- ssim ---------------------------]
                              |     ssim     |  ssim_separable
1 threads: ---------------------------------------------------
      cpu, 1, 128, 3x3        |      1280.5  |       2115.7   
      cpu, 1, 128, 11x11      |      3220.1  |       2467.5   
      cpu, 1, 128, 25x25      |     12284.1  |       3547.4   
      cpu, 1, 128, 51x51      |     47879.9  |       5389.0   
      cpu, 1, 128, 111x111    |    246027.8  |       9992.3   
      cpu, 1, 256, 3x3        |      3291.9  |       5466.6   
      cpu, 1, 256, 11x11      |     11032.3  |       6921.2   
      cpu, 1, 256, 25x25      |     47252.4  |      11015.9   
      cpu, 1, 256, 51x51      |    188903.8  |      18130.9   
      cpu, 1, 256, 111x111    |    998907.7  |      38439.7   
      cpu, 1, 512, 3x3        |     25763.3  |      34568.9   
      cpu, 1, 512, 11x11      |     57373.8  |      41282.8   
      cpu, 1, 512, 25x25      |    202391.9  |      59122.5   
      cpu, 1, 512, 51x51      |    769014.6  |      92438.8   
      cpu, 1, 512, 111x111    |   4007487.1  |     162060.4   
      cpu, 1, 1024, 3x3       |    187399.5  |     288808.3   
      cpu, 1, 1024, 11x11     |    308552.6  |     332325.1   
      cpu, 1, 1024, 25x25     |    889529.0  |     393250.9   
      cpu, 1, 1024, 51x51     |   3187592.3  |     496850.0   
      cpu, 1, 1024, 111x111   |  15927975.2  |     768873.2   
      cpu, 3, 128, 3x3        |      2607.6  |       4241.7   
      cpu, 3, 128, 11x11      |      8498.8  |       5256.3   
      cpu, 3, 128, 25x25      |     35689.4  |       8499.2   
      cpu, 3, 128, 51x51      |    142065.9  |      13991.5   
      cpu, 3, 128, 111x111    |    748177.1  |      27687.6   
      cpu, 3, 256, 3x3        |     10746.6  |      17170.3   
      cpu, 3, 256, 11x11      |     33625.1  |      20537.7   
      cpu, 3, 256, 25x25      |    142636.3  |      32972.7   
      cpu, 3, 256, 51x51      |    578064.8  |      55177.1   
      cpu, 3, 256, 111x111    |   2997392.4  |     116703.2   
      cpu, 3, 512, 3x3        |     75232.3  |     127123.3   
      cpu, 3, 512, 11x11      |    160919.3  |     139431.3   
      cpu, 3, 512, 25x25      |    624840.1  |     207684.5   
      cpu, 3, 512, 51x51      |   2294788.3  |     296566.2   
      cpu, 3, 512, 111x111    |  11883141.5  |     516269.7   
      cpu, 3, 1024, 3x3       |    614233.6  |     988069.4   
      cpu, 3, 1024, 11x11     |    989627.6  |    1068152.1   
      cpu, 3, 1024, 25x25     |   2728403.6  |    1252902.2   
      cpu, 3, 1024, 51x51     |   9526034.5  |    1650768.2   
      cpu, 3, 1024, 111x111   |  49049597.6  |    2533918.6   
      cuda, 1, 128, 3x3       |       553.7  |        904.7   
      cuda, 1, 128, 11x11     |       567.8  |        914.4   
      cuda, 1, 128, 25x25     |       803.8  |        907.1   
      cuda, 1, 128, 51x51     |      2035.2  |        913.8   
      cuda, 1, 128, 111x111   |      8572.4  |       1008.6   
      cuda, 1, 256, 3x3       |       570.1  |        915.6   
      cuda, 1, 256, 11x11     |       715.0  |        915.4   
      cuda, 1, 256, 25x25     |      1669.6  |        957.4   
      cuda, 1, 256, 51x51     |      5729.5  |       1118.1   
      cuda, 1, 256, 111x111   |     25760.3  |       1577.5   
      cuda, 1, 512, 3x3       |       824.1  |       1238.4   
      cuda, 1, 512, 11x11     |      1810.9  |       1500.8   
      cuda, 1, 512, 25x25     |      5546.4  |       1979.9   
      cuda, 1, 512, 51x51     |     20155.1  |       2889.5   
      cuda, 1, 512, 111x111   |     91134.0  |       5035.8   
      cuda, 1, 1024, 3x3      |      2758.6  |       3967.7   
      cuda, 1, 1024, 11x11    |      6648.9  |       5007.5   
      cuda, 1, 1024, 25x25    |     21440.3  |       6895.1   
      cuda, 1, 1024, 51x51    |     78729.4  |      10524.1   
      cuda, 1, 1024, 111x111  |    358589.0  |      18398.9   
      cuda, 3, 128, 3x3       |       569.7  |        911.2   
      cuda, 3, 128, 11x11     |       627.7  |        916.2   
      cuda, 3, 128, 25x25     |      1203.7  |        921.3   
      cuda, 3, 128, 51x51     |      3955.7  |       1020.2   
      cuda, 3, 128, 111x111   |     17500.6  |       1328.0   
      cuda, 3, 256, 3x3       |       664.3  |       1036.4   
      cuda, 3, 256, 11x11     |      1414.7  |       1236.6   
      cuda, 3, 256, 25x25     |      4270.7  |       1604.8   
      cuda, 3, 256, 51x51     |     15328.3  |       2288.9   
      cuda, 3, 256, 111x111   |     69503.1  |       4114.2   
      cuda, 3, 512, 3x3       |      2107.9  |       3048.7   
      cuda, 3, 512, 11x11     |      5025.2  |       3834.8   
      cuda, 3, 512, 25x25     |     16105.9  |       5259.7   
      cuda, 3, 512, 51x51     |     58879.1  |       7903.4   
      cuda, 3, 512, 111x111   |    268764.1  |      14404.1   
      cuda, 3, 1024, 3x3      |      7907.5  |      11256.7   
      cuda, 3, 1024, 11x11    |     19591.6  |      14380.6   
      cuda, 3, 1024, 25x25    |     63739.2  |      19996.0   
      cuda, 3, 1024, 51x51    |    234758.8  |      31051.9   
      cuda, 3, 1024, 111x111  |   1072391.6  |      54840.3   

Times are in microseconds (us).
```
